File: Eshop.Domain/Model/PCBuild.cs
using Eshop.Domain.Identity;
using Eshop.Domain.Relationships;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace Eshop.Domain.Model
{
    public class PCBuild : BaseEntity
    {
        // Relationships

        [JsonIgnore]
        public string? UserId { get; set; }
        [JsonIgnore]
        public EshopUser? User { get; set; }

        [JsonIgnore]
        public IList<ProductInPcBuild> Products { get; set; } = new List<ProductInPcBuild>();

        // Display properties

        [NotMapped]
        public ProductInPcBuild? Processor
        {
            get
            {
                return Products.FirstOrDefault(x => x.Key == PcBuildKeys.Processor);
            }
        }

        [NotMapped]
        public ProductInPcBuild? Ram
        {
            get
            {
                return Products.FirstOrDefault(x => x.Key == PcBuildKeys.RAM);
            }
        }

        [NotMapped]
        public ProductInPcBuild? Ssd
        {
            get
            {
                return Products.FirstOrDefault(x => x.Key == PcBuildKeys.SSD);
            }
        }

        [NotMapped]
        public ProductInPcBuild? Hdd
        {
            get
            {
                return Products.FirstOrDefault(x => x.Key == PcBuildKeys.HDD);
            }
        }

        [NotMapped]
        public ProductInPcBuild? GraphicsCard
        {
            get
            {
                return Products.FirstOrDefault(x => x.Key == PcBuildKeys.GraphicsCard);
            }
        }

        [NotMapped]
        public ProductInPcBuild? PcCase
        {
            get
            {
                return Products.FirstOrDefault(x => x.Key == PcBuildKeys.PcCase);
            }
        }

        [NotMapped]
        public ProductInPcBuild? PowerSupply
        {
            get
            {
                return Products.FirstOrDefault(x => x.Key == PcBuildKeys.PowerSupply);
            }
        }

        [NotMapped]
        public ProductInPcBuild? Motherboard
        {
            get
            {
                return Products.FirstOrDefault(x => x.Key == PcBuildKeys.Motherboard);
            }
        }

        [NotMapped]
        public double TotalPrice
        {
            get
            {
                return Products.Select(x => x.Price).Sum();
            }
        }

        // Compatibility

        public int MotherboardProccesorCompatibility
        {
            get
            {
                if (Motherboard == null) return 0;
                if (Processor == null) return 0;

                var motherBoardSocketVendor = Motherboard?.Product?.Tags.FirstOrDefault(x => x.Key == TagKeys.CPU_MANUFACTURER);
                var processorSocketVendor = Processor?.Product?.Tags.FirstOrDefault(x => x.Key == TagKeys.CPU_MANUFACTURER);
                if (motherBoardSocketVendor == null || processorSocketVendor == null) return 0;
                if (motherBoardSocketVendor.Value != processorSocketVendor.Value) return -1;

                var motherBoardSocket = Motherboard?.Product?.Tags.FirstOrDefault(x => x.Key == TagKeys.SOCKET);
                var processorSocket = Processor?.Product?.Tags.FirstOrDefault(x => x.Key == TagKeys.SOCKET);
                if (motherBoardSocket == null || processorSocket == null) return 0;
                if (motherBoardSocket.Value != processorSocket.Value) return -1;

                return 1;
            }
        }

        public int MotherboardRamCompatibility
        {
            get
            {
                if (Motherboard == null) return 0;
                if (Ram == null) return 0;

                var motherboardRamType = Motherboard?.Product?.Tags.FirstOrDefault(x => x.Key == TagKeys.RAM_TYPE);
                var ramType = Ram?.Product?.Tags.FirstOrDefault(x => x.Key == TagKeys.RAM_TYPE);
                if (motherboardRamType == null || ramType == null) return 0;
                if (ramType.Value != motherboardRamType.Value) return -1;

                return 1;
            }
        }

    }

    public static class PcBuildKeys
    {
        public const string Processor = "Processors";
        public const string RAM = "RAM";
        public const string SSD = "Solid State Drives";
        public const string HDD = "Hard Drives";
        public const string Motherboard = "Motherboards";
        public const string PcCase = "PP Cases";
        public const string GraphicsCard = "Graphics Cards";
        public const string PowerSupply = "Power Supplies";
    }
}

File: Eshop.Frontend/eshop.store/src/components/PCBuild/PCBuild.js
import React, { useState, useEffect } from "react";
import { useNavigate } from "react-router-dom";

import useAxiosPrivate from "../../Hooks/useAxiosPrivate";
import PCBuildApiService from "../../api/PCBuildApiService";
import PCBuildItem from "./PCBuildItem";

import "./style.css";

const PCBuild = () => {
  const axiosPrivate = useAxiosPrivate();
  const pcBuildApi = new PCBuildApiService(axiosPrivate);
  const navigate = useNavigate();

  const [pcBuild, setPcBuild] = useState({});
  const [render, setRender] = useState(0);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const fetchPcBuild = async () => {
      try {
        const response = await pcBuildApi.getUserPcBuild();
        setPcBuild(response.data);
        setLoading(false);
      } catch (error) {
        console.log(error);
      }
    };
    fetchPcBuild();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [render]);

  const getCompatible = () => {
    let htmlTagCompatibility = "<b class";

    if (pcBuild.motherboardProccesorCompatibility === -1) {
      return `${htmlTagCompatibility}="text-danger">X</b>`;
    }

    if (pcBuild.motherboardRamCompatibility === -1) {
      return `${htmlTagCompatibility}="text-danger">X</b>`;
    }

    return `${htmlTagCompatibility}="text-success">✓</b>`;
  };

  const handleOrderClick = async () => {
    try {
      await pcBuildApi.orderPc();
      navigate("/cart");
    } catch (error) {
      console.log(error);
    }
  };

  return loading ? (
    <h1 className="text-light">Loading...</h1>
  ) : (
    <div className="container mt-2">
      <table className="table table-bordered">
        <thead>
          <tr className="pc-build-text">
            <th scope="col">Type</th>
            <th scope="col">Image</th>
            <th scope="col">Name</th>
            <th scope="col">Price</th>
            <th scope="col">Quantity</th>
            <th scope="col">Select</th>
          </tr>
        </thead>
        <tbody>
          <PCBuildItem
            item={pcBuild.motherboard}
            type={"Motherboards"}
            pcBuild={pcBuild}
            setRender={setRender}
          />
          <PCBuildItem
            item={pcBuild.processor}
            type={"Processors"}
            pcBuild={pcBuild}
            setRender={setRender}
          />
          <PCBuildItem
            item={pcBuild.ram}
            type={"RAM"}
            pcBuild={pcBuild}
            setRender={setRender}
          />
          <PCBuildItem
            item={pcBuild.graphicsCard}
            type={"Graphics Cards"}
            pcBuild={pcBuild}
            setRender={setRender}
          />
          <PCBuildItem
            item={pcBuild.ssd}
            type={"Solid State Drives"}
            pcBuild={pcBuild}
            setRender={setRender}
          />
          <PCBuildItem
            item={pcBuild.hdd}
            type={"Hard Drives"}
            pcBuild={pcBuild}
            setRender={setRender}
          />
          <PCBuildItem
            item={pcBuild.powerSupply}
            type={"Power Supplies"}
            pcBuild={pcBuild}
            setRender={setRender}
          />
          <PCBuildItem
            item={pcBuild.pcCase}
            type={"PC Cases"}
            pcBuild={pcBuild}
            setRender={setRender}
          />
          <tr className="pc-build-text">
            <th colSpan="3">Total Price:</th>
            <td>{pcBuild.totalPrice}</td>
            <th>Compatibility:</th>
            <th>
              <strong
                dangerouslySetInnerHTML={{ __html: getCompatible() }}
              ></strong>
            </th>
          </tr>
        </tbody>
      </table>
      <div className="row mt-2">
        <div className="col-md-2">
          <button className="btn btn-primary w-100" onClick={handleOrderClick}>
            Order now!
          </button>
        </div>
      </div>
    </div>
  );
};

export default PCBuild;
